### PR TITLE
Fix Json issues: parse on form and tailrec deser

### PR DESCRIPTION
**Pull request by octonato**

### CODE REVIEW

**Review by ignasi35 — APPROVED** [review]

File: core/play-integration-test/src/it/scala/play/it/http/parsing/JacksonJsonBodyParserSpec.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    "parse a simple JSON body with custom Jackson json-read-features" in new WithApplication(
+      guiceBuilder =>
+        guiceBuilder.configure(
+          "akka.serialization.jackson.play.json-read-features.ALLOW_SINGLE_QUOTES" -> "true"
+        )
+    ) {
+
+      val configuration: Configuration = implicitly[Application].configuration
+      configuration.get[Boolean]("akka.serialization.jackson.play.json-read-features.ALLOW_SINGLE_QUOTES") must_== true
```
Comment (cchantep) [inline core/play-integration-test/src/it/scala/play/it/http/parsing/JacksonJsonBodyParserSpec.scala:60]: `must beTrue`

File: core/play-integration-test/src/it/scala/play/it/http/parsing/JacksonJsonBodyParserSpec.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+      val either: F.Either[Result, JsonNode] = parse("""{ 'field1':'value1' }""")
+      either.left.ifPresent(verboseFailure)
+      either.right.get().get("field1").asText() must_== "value1"
```
Comment (cchantep) [inline core/play-integration-test/src/it/scala/play/it/http/parsing/JacksonJsonBodyParserSpec.scala:64]: `must_===`

**Comment by octonato** [comment]
@Mergifyio backport master

File: core/play/src/main/scala/play/api/data/Form.scala [hunk 3]
```diff
@@ -977,3 +1064,6 @@ trait ObjectMapping {
     }
   }
 }
+
+case class FormJsonExpansionTooLarge(limit: Long)
+    extends RuntimeException(s"Binding form from JSON exceeds form expansion limit of $limit")
```
Comment (cchantep) [inline core/play/src/main/scala/play/api/data/Form.scala:1069]: `NoStackTrace` ?

**Comment by octonato** [comment]
Thanks for the review @cchantep.

I applied your suggestions and did the same to #10496 but didn't push yet. I will wait to get this one merged first to avoid overloading Travis. Travis have been quite slow, tired and moody recently.

**Comment by maxkremer** [comment]
This broke any custom deserializers we were using that made use of `JsonParser.readTree()`. 
The following stack trace is what we're experiencing:
```
aused by: com.fasterxml.jackson.databind.JsonMappingException: We should be reading map, something got wrong (through reference chain: java.lang.Object[][0]->storage.jooq.TenantRecord["channelClassifier"]->models.tenant.ChannelClassifier["channels"]->java.util.ArrayList[0]->models.tenant.Channel["channelRules"]->models.analytics.query.FilterExpression["rules"]->java.util.ArrayList[1])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:390)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:361)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:363)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:347)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187)
	at com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.deserialize(ObjectArrayDeserializer.java:214)
	at com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.deserialize(ObjectArrayDeserializer.java:24)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4593)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3548)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3516)
	at storage.multi.RestPlayImpl.lambda$getTenants$0(RestPlayImpl.java:48)
	... 10 common frames omitted
Caused by: java.lang.RuntimeException: We should be reading map, something got wrong
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:141)
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:56)
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:52)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4569)
	at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:2908)
	at com.fasterxml.jackson.core.JsonParser.readValueAsTree(JsonParser.java:2311)
	at models.serializers.FilterTermTypeDectectingDeserializer.deserialize(FilterTermTypeDectectingDeserializer.java:48)
	at models.serializers.FilterTermTypeDectectingDeserializer.deserialize(FilterTermTypeDectectingDeserializer.java:27)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:347)
	... 34 common frames omitted
```
It seems like the parser is in an inconsistent state due to consuming an extra token.

**Comment by mkurz** [comment]
@maxkremer So is this a bug in Play that needs to be fixed?
@octonato @ignasi35 WDYT?

**Comment by octonato** [comment]
Yep, seems to be a bug, but I wouldn't consider it a priority since it's an edge case and there is a workaround as suggested by @maxkremer. 

Ultimately, the issue will go away once we can move to Jackson 2.12 (as discussed in https://github.com/playframework/playframework/issues/10541)

**Comment by mkurz** [comment]
Thanks Renato!

**Pull request merged**